src/material/checkbox/checkbox.scss

```diff
@@ -1,71 +1,167 @@
+        selected-focus-icon-color: null,
+        unselected-focus-icon-color: null,
+      // MDC expects `.mdc-checkbox__ripple::before` to be the state layer, but we use
+      // `.mdc-checkbox__ripple` instead, so we emit the state layer slots ourselves.
+        unselected-hover-state-layer-opacity: null,
+        unselected-hover-state-layer-color: null,
+        unselected-focus-state-layer-opacity: null,
+        unselected-focus-state-layer-color: null,
+        unselected-pressed-state-layer-opacity: null,
+        unselected-pressed-state-layer-color: null,
+        selected-hover-state-layer-opacity: null,
+        selected-hover-state-layer-color: null,
+        selected-focus-state-layer-opacity: null,
+        selected-focus-state-layer-color: null,
+        selected-pressed-state-layer-opacity: null,
+        selected-pressed-state-layer-color: null,
+    )));
+
+    @include token-utils.use-tokens(tokens-mdc-checkbox.$prefix, $mdc-checkbox-slots) {
+      // MDC expects focus on .mdc-checkbox, but we focus the native element instead, so we need to
+      // emit a our own slot for the focus styles.
+      .mdc-checkbox__native-control:enabled:focus {
+        // Extra `:focus` included to achieve higher specificity than MDC's `:hover` style.
+        &:focus:not(:checked):not(:indeterminate) ~ .mdc-checkbox__background {
+          @include token-utils.create-token-slot(border-color, unselected-focus-icon-color);
+        }
+
+        &:checked,
+        &:indeterminate {
+          & ~ .mdc-checkbox__background {
+            @include token-utils.create-token-slot(border-color, selected-focus-icon-color);
+            @include token-utils.create-token-slot(background-color, selected-focus-icon-color);
+          }
+        }
+      }
 
-  // Avoids issues in some CSS grid layouts (see #25153).
-  position: relative;
-  -webkit-tap-highlight-color: transparent;
+      // MDC expects `.mdc-checkbox__ripple::before` to be the state layer, but we use
+      // `.mdc-checkbox__ripple` instead, so we emit the state layer slots ourselves.
+      &:hover {
+        .mdc-checkbox__ripple {
+          @include token-utils.create-token-slot(opacity, unselected-hover-state-layer-opacity);
+          @include token-utils.create-token-slot(
+                  background-color, unselected-hover-state-layer-color);
+        }
+
+        .mat-mdc-checkbox-ripple .mat-ripple-element {
+          @include token-utils.create-token-slot(
+                  background-color, unselected-hover-state-layer-color);
+        }
+      }
 
-  .mdc-checkbox {
-    // MDC theme styles also include structural styles so we have to include the theme at least
-    // once here. The values will be overwritten by our own theme file afterwards.
-    @include mdc-helpers.disable-mdc-fallback-declarations {
-      @include mdc-checkbox-theme.theme-styles(checkbox-private.$private-checkbox-theme-config);
-    }
+      .mdc-checkbox__native-control:focus {
```

Comment:
nit: it seems like the `& ~ .mdc-checkbox__ripple` and `& ~ .mat-mdc-checkbox-ripple .mat-ripple-element` are repeated a few times. Maybe we can clean this up a bit with a mixin?

Reply:
I did try to extract it, but I felt like it just made the code harder to reason about. I'll give it another try to see if I can come up with something that actually simplifies it instead of just making it harder to follow.

I think the real solution to this is to just do what MDC exepcts w.r.t. the state layer and stop trying to do our own thing. Then we can remove all this code and rely on the MDC mixin